doc/dev/haskell-idioms.md

```diff
@@ -0,0 +1,182 @@
+emulated with [GADT]s like so:
+```haskell
+{-# LANGUAGE GADTs #-}
+
+data Some f where
+  Some :: f x -> Some f  -- the type does not mention `x`
+```
+(`Some` is canonically available as [`Data.Parameterized.Some.Some`].) As
+defined above, `Some` has kind `forall k. (k -> Type) -> Type`. What if you want
+to existentially quantify over something of a different kind, e.g., you want to
+close over the first parameter of a datatype with two parameters? Then, you have
+to make your own GADT, and such types are conventionally named `Some*`.
+
+[first-class existential types]: https://github.com/ghc-proposals/ghc-proposals/pull/473
+
+## Faux `let`-bindings in type signatures
+
+Sometimes, a long type like `AVeryLongTypeName a (AnotherLongName b)` appears
+several times in one signature:
+```haskell
+foo ::
+  ( Eq (AVeryLongTypeName a (AnotherLongName b))
+  , Hashable (AVeryLongTypeName a (AnotherLongName b))
+  , Show (AVeryLongTypeName a (AnotherLongName b))
+  ) =>
+  AVeryLongTypeName a (AnotherLongName b) ->
+  AVeryLongTypeName a (AnotherLongName b) ->
+  ()
+foo _x _y = ()
+```
+One could define a type synonym
+```haskell
+type ShorterName a b = AVeryLongTypeName a (AnotherLongName b)
+```
+However, this forces readers of the code to know about (and remember) the
+synonym. Alternatively, one can use equality constraints to form a "faux
+`let`-binding" like so:
+```haskell
+foo ::
+  ( t ~ AVeryLongTypeName a (AnotherLongName b)
+  , Eq t
+  , Hashable t
+  , Show t
+  ) =>
+  t ->
+  t ->
+  ()
+foo _x _y = ()
+```
+
+## Pattern matching on `pure` values in `do` blocks
+
+[GADT] constructors can existentially quantify types.
+```haskell
+{-# LANGUAGE GADTs #-}
+
+data Some f where
+  Some :: f x -> Some f  -- the type does not mention `x`
+```
+(`Some` is canonically available as [`Data.Parameterized.Some.Some`].) To use
+a `Some`, you generally pattern match on it. This brings the captured variable
+into scope (though just as an unconstrained variable, so any code you write
+after the pattern match has to be polymorphic over it).
+```haskell
+wasItJust :: Some Maybe -> String
+wasItJust someMaybe =
+  case someMaybe of
+    -- Can use _x here, but have to be completely polymorphic over its type. Not
+    -- really much you *can* do with a value of an unknown type.
+    Some (Just _x) -> "it was Just"
+    Some Nothing -> "it was Nothing"
+
+main :: IO ()
+main = putStrLn (wasItJust (Some (Just "ok")))
+```
+Sometimes, we want to bring the variable into scope by pattern-matching on
+the constructor, but we'd rather not use a `case` statement, as it increases
+indentation (e.g., we might be pattern-matching on several `Some`s in a row). We
+can't do this with `let`, because the type variable is only in scope for the RHS
+of the `=`.
+```haskell
+wontWork :: Some Maybe -> IO ()
+wontWork someMaybe = do
+  let Some x = someMaybe
+  -- try something with x...
+  pure ()
```

Comment:
This example is perhaps not the most compelling, as you could simply match on `Some` in the function clause (i.e., `wontWork (Some x) = do ...`). To make it a bit more compelling, it would be worth introducing another function that returns a `Some Maybe` value, then call that function from `wontWork`.